035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,129 @@
+# Cluster Wide Volume Usage Quota Management
+
+## Current situation
+
+The [kafka-quota-plugin](https://github.com/strimzi/kafka-quotas-plugin) applies rate limits to connections to
+individual brokers, slowing producers down when consumed storage is over the soft limit (but below the hard limit) and
+effectively pausing publication when at or over the hard limit. This largely prevents out of disk scenarios when topics
+are replicated to all brokers thus there is largely similar disk usage between all brokers. Assuming all the brokers
+have similar disk usage levels they will all apply rate limits at similar times and levels this effectively gives
+cluster-wide out of storage protection.
+
+However, as clusters scale up the likelihood of even topic distribution drops. When topics are not evenly distributed it
+is possible for replication from broker `A` to broker `B` will cause broker `B` to consume all available disk space
+without triggering throttling of clients, as broker `A` disk usage is remains acceptable.
+
+[KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas) replication quotas are designed to
+manage the additional network load of migrating replicas between brokers rather and does not address client publication
+leading to out of disk conditions through replication.
+
+Currently, the quota-plugin considers the total quantity of storage and how much of that is consumed (
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect
+to [KIP-112](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the broker
+will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go-offline
+if all volumes are unavailable.
+
+Configuring the soft and hard limit thresholds is currently done in terms of bytes consumed, which forces users
+deploying the plugin to calculate a threshold for throttling and would require separate thresholds per volume to avoid
+out of space issues.
+
+## Proposal
+
+### Leverage kafka to distribute volume usage metrics throughout the cluster.
+
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID each instance of the plugin will be able
```

Comment:
It is possible to infer it from the [updateClusterMetadata](https://kafka.apache.org/30/javadoc/org/apache/kafka/server/quota/ClientQuotaCallback.html#updateClusterMetadata(org.apache.kafka.common.Cluster)) method on the ClientQuotaCallback interface.

However it might be simpler and more timely to inject it through configuration.